examples/encoder-decoder-coupling.lua

```diff
@@ -6,54 +6,107 @@ Example of "coupled" separate encoder and decoder networks, e.g. for sequence-to
+      end
+   end
 end
 
 --[[ Backward coupling: Copy decoder gradients to encoder LSTM ]]--
-local function backwardConnect(encLSTM, decLSTM)
-   encLSTM.userNextGradCell = nn.rnn.recursiveCopy(encLSTM.userNextGradCell, decLSTM.userGradPrevCell)
-   encLSTM.gradPrevOutput = nn.rnn.recursiveCopy(encLSTM.gradPrevOutput, decLSTM.userGradPrevOutput)
+function backwardConnect(enc, dec)
+   for i=1,#enc.lstmLayers do
+      if opt.useSeqLSTM then
+         enc.lstmLayers[i].userNextGradCell = dec.lstmLayers[i].userGradPrevCell
+         enc.lstmLayers[i].gradPrevOutput = dec.lstmLayers[i].userGradPrevOutput
+      else
+         enc.lstmLayers[i].userNextGradCell = nn.rnn.recursiveCopy(enc.lstmLayers[i].userNextGradCell, dec.lstmLayers[i].userGradPrevCell)
+         enc.lstmLayers[i].gradPrevOutput = nn.rnn.recursiveCopy(enc.lstmLayers[i].gradPrevOutput, dec.lstmLayers[i].userGradPrevOutput)
+      end
+   end
 end
 
 -- Encoder
 local enc = nn.Sequential()
-enc:add(nn.LookupTable(opt.vocabSize, opt.hiddenSize))
-enc:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
-local encLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize)
-enc:add(nn.Sequencer(encLSTM))
-enc:add(nn.SelectTable(-1))
+enc:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
+if not opt.useSeqLSTM then
+   enc:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
+end
+enc.lstmLayers = {}
+for i=1,opt.numLayers do
+   if opt.useSeqLSTM then
+      enc.lstmLayers[i] = nn.SeqLSTM(opt.hiddenSize, opt.hiddenSize)
+      enc.lstmLayers[i]:maskZero()
+      enc:add(enc.lstmLayers[i])
+   else
+      enc.lstmLayers[i] = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
+      enc:add(nn.Sequencer(enc.lstmLayers[i]))
+   end
+end
+if not opt.useSeqLSTM then
+   enc:add(nn.SelectTable(-1))
+end
 
 -- Decoder
 local dec = nn.Sequential()
-dec:add(nn.LookupTable(opt.vocabSize, opt.hiddenSize))
-dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
-local decLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize)
-dec:add(nn.Sequencer(decLSTM))
-dec:add(nn.Sequencer(nn.Linear(opt.hiddenSize, opt.vocabSize)))
-dec:add(nn.Sequencer(nn.LogSoftMax()))
-
-local criterion = nn.SequencerCriterion(nn.ClassNLLCriterion())
-
--- Some example data (batchsize = 2)
--- The input sentences to the encoder. 
-local encInSeq = torch.Tensor({{1,2,3},{3,2,1}})
--- The input sentences to the decoder. Label '5' represents the start of a sentence (GO).
-local decInSeq = torch.Tensor({{5,1,2,3,4},{5,4,3,2,1}})
--- The expected output from the decoder (it will return one character per time-step).
--- Label '6' represents the end of sentence (EOS).
-local decOutSeq = torch.Tensor({{1,2,3,4,6},{1,2,4,3,6}})
+dec:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
+if not opt.useSeqLSTM then
+   dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
```

Comment:
You can get rid of this altogether since Sequencer now also takes non table inputs.